File: user-service/logger.js
const amqp = require('amqplib');
const { v4: uuidv4 } = require('uuid');

const RABBITMQ_URL = process.env.RABBITMQ_URL || 'amqp://guest:guest@rabbitmq:5672/';
const EXCHANGE_NAME = 'logs';
const ROUTING_KEY = 'log';
const SERVICE_NAME = 'user-service';

let connection = null;
let channel = null;
let isConnecting = false;

// Initialize RabbitMQ connection
async function initializeLogger() {
  if (connection) return;
  if (isConnecting) {
    // Wait for connection to be established
    await new Promise(resolve => {
      const checkInterval = setInterval(() => {
        if (connection && channel) {
          clearInterval(checkInterval);
          resolve();
        }
      }, 100);
    });
    return;
  }

  isConnecting = true;
  try {
    connection = await amqp.connect(RABBITMQ_URL);
    channel = await connection.createChannel();
    
    // Ensure exchange exists
    await channel.assertExchange(EXCHANGE_NAME, 'direct', { durable: true });
    
    // Ensure queue exists
    await channel.assertQueue('logs_queue', { durable: true });
    
    // Bind queue to exchange
    await channel.bindQueue('logs_queue', EXCHANGE_NAME, ROUTING_KEY);
    
    console.log('[Logger] RabbitMQ connection established');
    console.log('[Logger] Exchange and queue configured successfully');
    isConnecting = false;
  } catch (err) {
    console.error('[Logger] Failed to connect to RabbitMQ:', err.message);
    isConnecting = false;
    // Retry connection after 5 seconds
    setTimeout(initializeLogger, 5000);
  }
}

// Format timestamp
function getTimestamp() {
  const now = new Date();
  const year = now.getFullYear();
  const month = String(now.getMonth() + 1).padStart(2, '0');
  const date = String(now.getDate()).padStart(2, '0');
  const hours = String(now.getHours()).padStart(2, '0');
  const minutes = String(now.getMinutes()).padStart(2, '0');
  const seconds = String(now.getSeconds()).padStart(2, '0');
  const ms = String(now.getMilliseconds()).padStart(3, '0');
  return `${year}-${month}-${date} ${hours}:${minutes}:${seconds},${ms}`;
}

// Format log message according to specification
function formatLog(timestamp, logType, url, correlationId, message) {
  return `${timestamp} ${logType} ${url} Correlation: ${correlationId} [${SERVICE_NAME}] - ${message}`;
}

// Send log to RabbitMQ
async function sendLogToRabbitMQ(logMessage) {
  if (!channel) {
    console.log('[Logger] RabbitMQ not connected yet, buffering log:', logMessage);
    return;
  }

  try {
    const logObject = {
      timestamp: new Date().toISOString(),
      service: SERVICE_NAME,
      message: logMessage
    };
    
    await channel.publish(
      EXCHANGE_NAME,
      ROUTING_KEY,
      Buffer.from(JSON.stringify(logObject))
    );
  } catch (err) {
    console.error('[Logger] Failed to send log to RabbitMQ:', err.message);
  }
}

// Internal logging function
function _log(logType, url, correlationId, message) {
  const timestamp = getTimestamp();
  const formattedLog = formatLog(timestamp, logType, url, correlationId, message);
  
  // Always log to console
  console.log(formattedLog);
  
  // Send to RabbitMQ asynchronously (don't block)
  sendLogToRabbitMQ(formattedLog).catch(err => {
    console.error('[Logger] Error in sendLogToRabbitMQ:', err.message);
  });
}

// Public logging functions
const logger = {
  info: (url, correlationId, message) => {
    _log('INFO', url, correlationId, message);
  },

  error: (url, correlationId, message) => {
    _log('ERROR', url, correlationId, message);
  },

  warn: (url, correlationId, message) => {
    _log('WARN', url, correlationId, message);
  },

  debug: (url, correlationId, message) => {
    _log('DEBUG', url, correlationId, message);
  }
};

// Graceful shutdown
async function closeLogger() {
  try {
    if (channel) await channel.close();
    if (connection) await connection.close();
    console.log('[Logger] RabbitMQ connection closed');
  } catch (err) {
    console.error('[Logger] Error closing RabbitMQ connection:', err.message);
  }
}

module.exports = {
  initializeLogger,
  logger,
  closeLogger
};

File: planner-service/logger.py
import json
import pika
import uuid
from datetime import datetime

RABBITMQ_URL = 'amqp://guest:guest@rabbitmq:5672/'
EXCHANGE_NAME = 'logs'
ROUTING_KEY = 'log'
SERVICE_NAME = 'planner-service'

connection = None
channel = None
is_connecting = False


def get_timestamp():
    """Format timestamp as specified"""
    now = datetime.now()
    return now.strftime('%Y-%m-%d %H:%M:%S,') + f'{now.microsecond // 1000:03d}'


def format_log(timestamp, log_type, url, correlation_id, message):
    """Format log message according to specification"""
    return f'{timestamp} {log_type} {url} Correlation: {correlation_id} [{SERVICE_NAME}] - {message}'


def send_log_to_rabbitmq(log_message):
    """Send log to RabbitMQ"""
    global channel
    
    if channel is None:
        print(f'[Logger] RabbitMQ not connected, buffering log: {log_message}')
        return
    
    try:
        log_object = {
            'timestamp': datetime.utcnow().isoformat() + 'Z',
            'service': SERVICE_NAME,
            'message': log_message
        }
        
        channel.basic_publish(
            exchange=EXCHANGE_NAME,
            routing_key=ROUTING_KEY,
            body=json.dumps(log_object)
        )
    except Exception as e:
        print(f'[Logger] Failed to send log to RabbitMQ: {str(e)}')


def _log(log_type, url, correlation_id, message):
    """Internal logging function"""
    timestamp = get_timestamp()
    formatted_log = format_log(timestamp, log_type, url, correlation_id, message)
    
    # Always log to console
    print(formatted_log)
    
    # Send to RabbitMQ asynchronously
    try:
        send_log_to_rabbitmq(formatted_log)
    except Exception as e:
        print(f'[Logger] Error in sendLogToRabbitMQ: {str(e)}')


class Logger:
    """Logger class with methods for different log levels"""
    
    @staticmethod
    def info(url, correlation_id, message):
        _log('INFO', url, correlation_id, message)
    
    @staticmethod
    def error(url, correlation_id, message):
        _log('ERROR', url, correlation_id, message)
    
    @staticmethod
    def warn(url, correlation_id, message):
        _log('WARN', url, correlation_id, message)
    
    @staticmethod
    def debug(url, correlation_id, message):
        _log('DEBUG', url, correlation_id, message)


# Export convenience functions
def log_info(url, correlation_id, message):
    """Log info message"""
    _log('INFO', url, correlation_id, message)


def log_error(url, correlation_id, message):
    """Log error message"""
    _log('ERROR', url, correlation_id, message)


def log_warn(url, correlation_id, message):
    """Log warning message"""
    _log('WARN', url, correlation_id, message)


def log_debug(url, correlation_id, message):
    """Log debug message"""
    _log('DEBUG', url, correlation_id, message)


def initialize_logger():
    """Initialize RabbitMQ connection"""
    global connection, channel, is_connecting
    
    if connection is not None:
        return
    
    if is_connecting:
        # Wait for connection
        import time
        while connection is None:
            time.sleep(0.1)
        return
    
    is_connecting = True
    try:
        connection = pika.BlockingConnection(
            pika.URLParameters(RABBITMQ_URL)
        )
        channel = connection.channel()
        
        # Ensure exchange exists
        channel.exchange_declare(
            exchange=EXCHANGE_NAME,
            exchange_type='direct',
            durable=True
        )
        
        # Ensure queue exists
        channel.queue_declare(
            queue='logs_queue',
            durable=True
        )
        
        # Bind queue to exchange
        channel.queue_bind(
            exchange=EXCHANGE_NAME,
            queue='logs_queue',
            routing_key=ROUTING_KEY
        )
        
        print('[Logger] RabbitMQ connection established')
        print('[Logger] Exchange and queue configured successfully')
        is_connecting = False
    except Exception as e:
        print(f'[Logger] Failed to connect to RabbitMQ: {str(e)}')
        is_connecting = False
        # Retry connection after 5 seconds
        import time
        import threading
        def retry():
            time.sleep(5)
            initialize_logger()
        threading.Thread(target=retry, daemon=True).start()


def close_logger():
    """Close RabbitMQ connection"""
    global connection, channel
    
    try:
        if channel is not None:
            channel.close()
        if connection is not None:
            connection.close()
        print('[Logger] RabbitMQ connection closed')
    except Exception as e:
        print(f'[Logger] Error closing RabbitMQ connection: {str(e)}')


logger = Logger()
